### PR TITLE
Tuple element names CDI refactoring

### PR DESCRIPTION
**Pull request by tmat**
Move tuple element names encoding to shared code and avoid wrapping values in TypedConstants that need to be unwrapped again when encoding CDI.

### CODE REVIEW

**Comment by tmat** [comment]
@dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs [hunk 1]
```diff
@@ -594,28 +594,30 @@ private static bool AddNames(TypeSymbol type, ArrayBuilder<string> namesBuilder)
...
                 Debug.Assert(flagsBuilder.Any());
                 Debug.Assert(flagsBuilder.Contains(true));
 
-                var constantsBuilder = ArrayBuilder<TypedConstant>.GetInstance(flagsBuilder.Count);
-                foreach (bool flag in flagsBuilder)
-                {
-                    constantsBuilder.Add(new TypedConstant(booleanType, TypedConstantKind.Primitive, flag));
-                }
-
+                var result = flagsBuilder.SelectAsArray((constantType, flag) => new TypedConstant(booleanType, TypedConstantKind.Primitive, constantType), booleanType);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs]: `SelectAsArray((flag, constantType) => new TypedConstant(constantType, TypedConstantKind.Primitive, flag), booleanType)`

**Comment by cston** [comment]
LGTM other than swapped parameters in `SelectAsArray`.

File: src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs (hunk 1 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs]: Ah, yes. Thanks

**Comment by tmat** [comment]
test windows_release_vs-integration_prtest please

**Comment by tmat** [comment]
test microbuild_prtest please

**Comment by tmat** [comment]
@dotnet/roslyn-compiler Ping.

File: src/Dependencies/CodeAnalysis.Metadata/CustomDebugInfoEncoder.cs [hunk 2]
```diff
@@ -143,20 +143,59 @@ public void AddDynamicLocals(IReadOnlyCollection<(string LocalName, byte[] Flags
...
-                        builder.WriteBytes(flags);
-                        builder.WriteInt32(count);
-                        builder.WriteInt32(slotIndex);
-                        builder.WriteUTF16(name);
-                        builder.WriteBytes(0, sizeof(char) * (IdentifierSize - name.Length));
+                        Debug.Assert(info.Flags.Length <= DynamicAttributeSize);
+                        Debug.Assert(info.LocalName.Length <= IdentifierSize);
+
+                        builder.WriteBytes(info.Flags);
+                        builder.WriteBytes(0, sizeof(byte) * (DynamicAttributeSize - info.Flags.Length));
```
Comment (jcouv) [inline src/Dependencies/CodeAnalysis.Metadata/CustomDebugInfoEncoder.cs]: I didn't understand this new output. Looks like we were not padding with zeros. Was that a bug?

Comment (tmat) [inline src/Dependencies/CodeAnalysis.Metadata/CustomDebugInfoEncoder.cs]: No, previously the method assumed the info.Flags array is itself padded with 0s to 64 elements. Now we relax this requirement on this array and the method fills in the 0s if there is less than 64 elements in the array.

File: src/Dependencies/CodeAnalysis.Metadata/CustomDebugInfoEncoder.cs [hunk 3]
```diff
@@ -143,20 +143,59 @@ public void AddDynamicLocals(IReadOnlyCollection<(string LocalName, byte[] Flags
...
+                        // Constants have slot index -1 and scope specified,
+                        // variables have a slot index specified and no scope.
+                        Debug.Assert((info.SlotIndex == -1) ^ (info.ScopeStart == 0 && info.ScopeEnd == 0));
+
+                        builder.WriteInt32(info.Names.Length);
+                        foreach (var name in info.Names)
+                        {
+                            if (name != null)
+                            {
+                                builder.WriteUTF8(name);
```
Comment (jcouv) [inline src/Dependencies/CodeAnalysis.Metadata/CustomDebugInfoEncoder.cs]: Is the behavior for `null` name seems different than before (`builder.WriteUTF8((string)tupleElementName.Value ?? string.Empty);`)?

Comment (tmat) [inline src/Dependencies/CodeAnalysis.Metadata/CustomDebugInfoEncoder.cs]: It's the same. Empty string in UTF8 is encoded as empty sequence of bytes. Which is the same as now writing the name out.

File: src/Test/Utilities/Portable/Assert/AssertXml.cs [hunk 4]
```diff
@@ -24,12 +24,12 @@ public static class AssertXml
     {
         public static void Equal(string expected, string actual)
         {
-            Equal(XElement.Parse(expected), XElement.Parse(actual), null, 0, expectedIsXmlLiteral: true);
+            Equal(XElement.Parse(expected), XElement.Parse(actual), null, null, 0, expectedIsXmlLiteral: true);
```
Comment (jcouv) [inline src/Test/Utilities/Portable/Assert/AssertXml.cs]: Consider naming parameters.

File: src/Test/Utilities/Portable/Assert/AssertXml.cs [hunk 5]
```diff
@@ -38,14 +38,15 @@ public static void Equal(XElement expected, XElement actual)
...
+            string messagePrefix,
             string expectedValueSourcePath,
             int expectedValueSourceLine,
             bool expectedIsXmlLiteral)
         {
             Tuple<XElement, XElement> firstMismatch;
             if (!CheckEqual(expectedRoot, actualRoot, ShallowElementComparer.Instance, out firstMismatch))
             {
-                Assert.True(false, GetAssertText(GetXmlString(expectedRoot, expectedIsXmlLiteral), GetXmlString(actualRoot, expectedIsXmlLiteral), expectedRoot, firstMismatch, expectedValueSourcePath, expectedValueSourceLine, expectedIsXmlLiteral));
+                Assert.True(false, messagePrefix + GetAssertText(GetXmlString(expectedRoot, expectedIsXmlLiteral), GetXmlString(actualRoot, expectedIsXmlLiteral), expectedRoot, firstMismatch, expectedValueSourcePath, expectedValueSourceLine, expectedIsXmlLiteral));
```
Comment (jcouv) [inline src/Test/Utilities/Portable/Assert/AssertXml.cs]: Please break this line.

File: src/Compilers/CSharp/Test/Emit/PDB/PDBConstantTests.cs [hunk 6]
```diff
@@ -14,7 +15,244 @@ public class PDBConstantTests : CSharpTestBase
         private CultureInfo _testCulture = new CultureInfo("en-US");
 
         [Fact]
-        public void TestSimpleLocalConstants()
+        public void StringsWithSurrogateChar()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/PDB/PDBConstantTests.cs]: Seems like those tests are unrelated to this PR (CDI refactoring for dynamic and tuple names).

Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/PDB/PDBConstantTests.cs]: These are just moved from a different file.

File: src/Compilers/Core/CodeAnalysisTest/Emit/CustomDebugInfoTests.cs [hunk 7]
```diff
@@ -587,6 +587,62 @@ public void DynamicLocals()
...
 
         [Fact]
+        public void TupleElementNames()
+        {
+            var builder = new BlobBuilder();
+            var cdiEncoder = new CustomDebugInfoEncoder(builder);
+
+            cdiEncoder.AddTupleElementNames(new[]
+            {
+                ("a", 1, 0, 0, ImmutableArray.Create("e")),
```
Comment (jcouv) [inline src/Compilers/Core/CodeAnalysisTest/Emit/CustomDebugInfoTests.cs]: It would be good to name those elements for readability. But then again, the names won't be verified :-(
